### ISSUE TITLE
[R] mx.nd.concat is expecting a MXNDArray instead of a list of MXNDArray

### ISSUE THREAD

**Issue by jeremiedb** [dmlc/mxnet#6650]
I think there might be a issue with the expected format of the `data` argument in `mx.nd.concat`. An error is returned if a list of MXNDArray is passed as argument mentionning:  

```
> x1 <- mx.nd.array(array(1:24, dim=c(4,6)))
> x2 <- mx.nd.array(array(1:24, dim=c(4,6)))
> x3 <- mx.nd.concat(data = c(x1, x2), num_args = 2, dim = 1)
Error in mx.nd.concat(data = c(x1, x2), num_args = 2, dim = 1) : 
  Expecting an external pointer: [type=list].
```

The equivalent code using `mx.symbol.concat` works fine.

**Comment by jeremiedb** [comment]
Actually it's `mx.symbol.Concat` that works properly, `mx.symbol.concat` is also causing issue: 

```
> x1 <- mx.symbol.Variable("data1")
> x2 <- mx.symbol.Variable("data2")
> y <- mx.symbol.concat(data=c(x1, x2), num.args = 2)
Error in mx.varg.symbol.concat(list(...)) : 
  ./base.h:271: RCheck failed: TYPEOF(val) == INTSXP || TYPEOF(val) == REALSXP Only accept integer vectors or simple types
```

Sorry for the confusion!

**Comment by jeremiedb** [comment]
I just recompiled and still cannot make the mx.nd.concat or mx.nd.Concat work. Did I miss something obvious? Thanks!

**Comment by thirdwing** [comment]
I can confirm this error and will fix it soon.

**Comment by thirdwing** [comment]
@jeremiedb I think this is what we expect, right?

```r
> library(mxnet)
> x1 <- mx.nd.array(array(1:24, dim=c(4,6)))
> x2 <- mx.nd.array(array(1:24, dim=c(4,6)))
> x3 <- mx.nd.concat(data = c(x1, x2), num_args = 2, dim = 1)
> x1
     [,1] [,2] [,3] [,4] [,5] [,6]
[1,]    1    5    9   13   17   21
[2,]    2    6   10   14   18   22
[3,]    3    7   11   15   19   23
[4,]    4    8   12   16   20   24
> x3
     [,1] [,2] [,3] [,4] [,5] [,6]
[1,]    1    5    9   13   17   21
[2,]    2    6   10   14   18   22
[3,]    3    7   11   15   19   23
[4,]    4    8   12   16   20   24
[5,]    1    5    9   13   17   21
[6,]    2    6   10   14   18   22
[7,]    3    7   11   15   19   23
[8,]    4    8   12   16   20   24
> class(x3)
[1] "MXNDArray"
> s1 <- mx.symbol.Variable("data1")
> s2 <- mx.symbol.Variable("data2")
> s3 <- mx.symbol.concat(data=c(s1, s2), num.args = 2)
> s3
C++ object <0x62dc200> of class 'MXSymbol' <0x4fe63f0>
> class(s3)
[1] "Rcpp_MXSymbol"
attr(,"package")
[1] "mxnet"
```

**Comment by jeremiedb** [comment]
Yes! This looks like the expected behavior. 

Also, I just realized that it might be the same type of check that needs to be performed in order to fix #5296. When running on multiple devices, I guess the kvstore$push returns a list rather than MXNDArray.
